### DISCUSSION TITLE
How do you unregister fields when they get unmounted?

### DISCUSSION THREAD

**Discussion by diegohaz** [react-hook-form/react-hook-form#3009] [Q&A] [answered]
Hi! I was working on something using a similar approach to `ref={register}` and I got stuck on this: how can I unregister the element if it gets unmounted if the `ref` parameter will be `null`?

```js
function register(ref) {
  if (ref === null) {
    // How to unregister the element if I don't have the ref anymore?
    // refs.current.delete(...)
  } else {
    refs.current.add(ref);
  }
}
```

I see that `react-hook-form` solves this. I wonder how?

**Comment by bluebill1049** [comment, marked as answer]
hey @diegohaz Awesome post on the other day with TS 4.1.0 (string literal). we have two primary ways to register an input. 
`ref={register}` and `Controller`, `Controller` is pretty easy with `useEffect`, with ref register: we are using `mutationObserver` to detect an input gets removed from the DOM, if it doesn't get re-mounted again with the same `name`, then we assure this is truly unmounted. hope this helps.

**Reply by diegohaz** [reply]
Ohhhh, this helps a lot! Thanks! Awesome work on this project! ❤️
